### PR TITLE
Upgrade to TypeScript 3.9

### LINKED ISSUE #10855: webpack has compile errors with Typescript 3.9
# Bug report

**What is the current behavior?**

webpack currently compiles with 3.8.3, but if you use `typescript@rc` instead (3.9.1-rc), you get the following 9 errors:

```
lib/ExternalModule.js(305,28): error TS2554: Expected 1 arguments, but got 0.
lib/Stats.js(68,31): error TS2554: Expected 3 arguments, but got 2.
lib/library/SystemLibraryPlugin.js(127,35): error TS2554: Expected 1 arguments, but got 0.
lib/optimize/ConcatenatedModule.js(205,32): error TS2554: Expected 1 arguments, but got 0.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(36,15): error TS2304: Cannot find name 'Chunk'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(37,15): error TS2304: Cannot find name 'DependencyTemplates'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(38,15): error TS2304: Cannot find name 'RuntimeTemplate'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(39,15): error TS2304: Cannot find name 'ModuleGraph'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(40,15): error TS2304: Cannot find name 'ChunkGraph'.
```

If you compile with the version from master, you get 12 additional errors:

```
lib/AbstractMethodError.js(26,16): error TS2532: Object is possibly 'undefined'.
lib/ContextModule.js(333,29): error TS2345: Argument of type 'Error' is not assignable to parameter of type 'WebpackError'.
  Type 'Error' is missing the following properties from type 'WebpackError': details, module, loc, hideStack, and 4 more.
lib/ExternalModule.js(305,28): error TS2554: Expected 1 arguments, but got 0.
lib/MainTemplate.js(40,29): error TS1016: A required parameter cannot follow an optional parameter.
lib/ModuleGraphConnection.js(23,3): error TS1016: A required parameter cannot follow an optional parameter.
lib/Stats.js(68,31): error TS2554: Expected 3 arguments, but got 2.
lib/config/normalization.js(27,30): error TS1016: A required parameter cannot follow an optional parameter.
lib/config/normalization.js(37,38): error TS1016: A required parameter cannot follow an optional parameter.
lib/config/normalization.js(47,29): error TS1016: A required parameter cannot follow an optional parameter.
lib/config/normalization.js(56,37): error TS1016: A required parameter cannot follow an optional parameter.
lib/library/SystemLibraryPlugin.js(127,35): error TS2554: Expected 1 arguments, but got 0.
lib/library/UmdLibraryPlugin.js(38,31): error TS1016: A required parameter cannot follow an optional parameter.
lib/optimize/ConcatenatedModule.js(205,32): error TS2554: Expected 1 arguments, but got 0.
lib/util/fs.js(57,23): error TS1016: A required parameter cannot follow an optional parameter.
lib/util/fs.js(78,19): error TS1016: A required parameter cannot follow an optional parameter.
lib/util/fs.js(98,22): error TS1016: A required parameter cannot follow an optional parameter.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(36,15): error TS2304: Cannot find name 'Chunk'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(37,15): error TS2304: Cannot find name 'DependencyTemplates'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(38,15): error TS2304: Cannot find name 'RuntimeTemplate'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(39,15): error TS2304: Cannot find name 'ModuleGraph'.
lib/wasm-async/AsyncWebAssemblyModulesPlugin.js(40,15): error TS2304: Cannot find name 'ChunkGraph'.
```

That's because typescript is now stricter about javascript marking optional properties. They have to be marked, and occur before non-optional properties.
The wasm-async errors were there all along, but we forgot to make the compiler check otherwise-unused properties of a `@typedef` before 3.9.

`typescript@next` also uses control flow on `this.x` property assignments in constructors to figure out their type. This gives better types (like in ContextModule), but also doesn't know about side effects (like in AbstractMethodError).

**If the current behavior is a bug, please provide the steps to reproduce.**

```sh
$ npm install -g typescript@rc # or typescript@next
$ tsc
```

This is mostly a heads up for the new version. I'll create a PR if I can fix the problems easily.

### PR DESCRIPTION
**Pull request by mohsen1**
Update to TypeScript 3.9 beta. 

> We should wait for 3.9 to be released before merging this. 

This PR adds some types to fix newly found type errors. I've made a few assumptions writing those types. Please correct me if I made a mistake making those assumptions 

## To do

* [x] Replace `@ts-ignore` with `@ts-expect-error`
* [x] Enable rule lint to disallow `@ts-ignore`

Fixes #10855 

**What kind of change does this PR introduce?**
Dependency update

**Did you add tests for your changes?**
NA

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**
Nothing

### CODE REVIEW

File: lib/WebpackOptionsApply.js [hunk 1]
```diff
@@ -588,7 +588,7 @@ class WebpackOptionsApply extends OptionsApply {
 					break;
 				}
 				default:
-					// @ts-ignore never is expected here
+					// TODO options.cache.type type is : "memory" | "filesystem"
```
Comment (mohsen1) [inline lib/WebpackOptionsApply.js]: @sokra Are you expecting `type` to be `never` here?

Comment (sokra) [reply lib/WebpackOptionsApply.js]: yes, memory and filesystem is handled in the switch cases above.

Comment (mohsen1) [reply lib/WebpackOptionsApply.js]: Ohhh it took me a while to figure this out. So `cacheOptions.type` is indeed `never`. Here we're using `options.cache.type` which its type is not being narrowed with the switch statement. TypeScript id working as expected. I'll remove the comment.

File: lib/WebpackOptionsApply.js [hunk 2]
```diff
@@ -588,7 +588,6 @@ class WebpackOptionsApply extends OptionsApply {
 					break;
 				}
 				default:
-					// @ts-ignore never is expected here
 					throw new Error(`Unknown cache type ${options.cache.type}`);
```
Comment (sokra) [inline lib/WebpackOptionsApply.js:591]:
```suggestion
					throw new Error(`Unknown cache type ${cacheOptions.type}`);
```

So you can use expect-error

**Comment by sokra** [comment]
@sandersn I think there is a small bug in this typescript release which affects our type generation.

There seem to be some deduplicaton of union/intersection types. But this seems to be a bit too aggressive, which causes an identical but otherwise unrelated type to have an `aliasSymbol` pointing to the previous symbol. Our type generation uses `aliasSymbol` to give types names. Now all identical types are merged into an unrelated name inferred from the `aliasSymbol`.

**Comment by sandersn** [comment]
@weswigham do you know some tricks to avoid alias de-duplication?

**Comment by weswigham** [comment]
It's entirely unavoidable for unions of primitives. Now, if you explicitly write out the names yourself, well, we use what you wrote...

**Comment by sokra** [comment]
Thanks

**Pull request merged**